californium-core/src/main/java/org/eclipse/californium/core/network/UdpMatcher.java

```diff
@@ -210,13 +209,19 @@ public void receiveRequest(final Request request, final EndpointReceiver receive
 		boolean duplicate = previous != null;
 
 		if (duplicate) {
-			// assuming addresses changing, request could not only be deduplicated by their address and mid.
-			EndpointContext sourceContext = request.getSourceContext();
-			Request previousRequest = previous.getRequest();
-			EndpointContext previousSourceContext = previousRequest.getSourceContext();
-			// the previous response would be send with its previous context using 
-			// the current request context as connection context
-			duplicate = endpointContextMatcher.isToBeSent(previousSourceContext, sourceContext);
+			// assuming addresses changing, request could not be
+			// deduplicated only by their address and MID.
+			if (previous.getOrigin() != Origin.REMOTE) {
+				// accidently match MID of response
+				duplicate = false;
```

Comment:
If this happens that means that device does not respect the specification ? 

Should we have this kind of "accident" for responce deduplication ?

Reply:
No. It mean's, the device and the server use both coap-client and coap-server roles. while one device send CON notifies, a next device, assigned to that address sends request.

Reply:
I didn't understand this at all. 

> while one device send CON notifies, a next device, assigned to that address sends request.

So the test would generate an incorrect behavior (false negative) if device does not change its address but reuse an MID between a CON response(e.g. notify) and a request ? Should we rather use echange.getOrigin to know if we should check source or destination endpoint context ?

(:thinking: if peer identity

Reply:
Next try:

device 1, ip1, is observed by the cloud server, and therefore sends CON notifies, which are considered to be deduplicated.

device 2, gets mapped to ip1 within the 247s since the last CON notify, and sends a request. Randomly (1/64k) it chooses the MID of the CON notify with the pending deduplication. 

Nothing is wrong, except, that the NAT reassigned the ip address before the 247s expired. That NAT's don't reassigng within that time is a requirement from LWM2M. And if every system complies to that, no special deduplication would be required. But, in my experience, it's hard to ensure that. And therefore I mitigated it in order not to fail the request.

> (thinking if peer identity is used all of this tests are obsolete, right ?)
  
Hope so, but I didn't spend too much time into verify that.

Reply:
Just to mention:

that errors/collisions occurred during test over a couple of hours. So this will rarely hit in the reality. But if, this PR mitigates the effect.

Reply:
That's what I was understanding (the second time).

What I tried to say is : 

With your PR : 
device 1 sends CON notify with a given MID.
device 1(same device, no peer address changing)  sends a request with the same MID. (this device acts wrongly)
We will not deduplicate as before as we will says this is not a duplicate.

So I asked my self, if this behavior change is correct.
If not, we should use exchange.getOrigin to know which endpoint context to use to check if peer changed.

Reply:
FMPOV, RFC7252 considers only valid MID usage, means, if the device sends a different message with the same MID within the 247s, I don't see, that this is specified in a useful way. Therefore I would not care too much about misbehaving devices and instead focus on unaware hits by address changes.

Reply:
Should we add a enpoint context check for that as well?
using the source context of the current response instead of the current request?
I don't think, this makes it better, but if you prefer it ...

Reply:
My experience is : "world is full of devices which does not respect the specification correctly"

How this happened ? My understanding is some developers don't really understand specification they are using. They code then test against a server, if it works that's OK and devices are shipped.  
One days we fix a too flexible behavior at server side, and boom this breaks some devices which work only with this flexible/out of spec behavior. 

The idea : 
 - If server never allowed this devices will never be shipped with this bugs.
 - In another way, a server should tolerate bad citizen devices...  

That's a common dilemma between be strict or flexible with input...

My current position is probably that the best solution is :
 - be strict and add flexibility only when needed and in a configurable way and log when this flexibility is used.

So here, I don't like so much the idea to let devices not respect the MID usage transparently.

> Therefore I would not care too much about misbehaving devices and instead focus on unaware hits by address changes.

So I care about both :)

Reply:
> I don't think, this makes it better, but if you prefer it ...

I prefer. You think this doesn't make it better because you think this useless ? or there is something else ?

Reply:
I'm lost.

I don't understand, what you assume to be the way to process in this case. Maybe you specify that and best, test, if the current implementation before that PR matches your expectation.

Reply:
> what you assume to be the way to process in this case

To know if a message is a duplicate, we use MID value + "identity/endpoint context".

So for example with : 

Case A)
device1 send a CON notify with MID=a
device1 send a request with MID=a

Case B)
device1 send a GET request with MID=b
device1 send a PUT request with MID=b

The 2nd message should be considered as a duplicate of the first one. (I think that was the case before we introduce the changing address/endpoint context check)

But now case A) and B) is not handled in a same way.
For A) 2nd request is no more considered as a duplicate and replace the notify.
For B) 2nd request is a duplicate.

Is it more clear ?

Reply:
These are not the details I expect!

Case a)
device 1 send CON (Notify or response) with MID=a (e.g. token=abcd)
=> device 1 gets an ACK with MID=a back.

device 1 send a CON Get with MID=a (token=efgh).

What do you expect now to be sent to the device?
ACK with MID=a and let the request timeout (with NSTARt-1 that blocks for 247s).
Or send the CON notify (response?) MID=a token=abcd back to the device (maybe that's the current behavior, but I haven tested it)?
Or ???

So, what do you prefer?

Reply:
This is what I expected : ACK with MID=a and let the request timeout.

And reading the code this is probably not what it will happen, so we can let the code like this...
(I will test it on my side just to know)

Reply:
> like this ...

What does this ... mean? This PR?

Reply:
And just to say:
What do you intend for your case B, PUT?
Only ACK? Piggyback response of the GET?

From my experience, starting with my issue #30, it cost you hours and hours to analyze a communication, where the message returned matches parts of your expectation and others not. And, if the MID usage is doubt, there is close to nothing you can do.

Reply:
> What does this ... mean? This PR?

I wanted say : "this PR"
But regarding my analyze below I think this is maybe better to add the test source endpoint test. (but no strong opinion)

> What do you intend for your case B, PUT ?

Same ACK as GET.

------------------------------------------------------------
Just in case this interest you :

I create a dummy test in ObserveClientSideTest (+ adding a coap server to clientendpoint)
```java
@Test
public void testNotifyRequestSameMID() throws Exception {
	System.out.println("Observe with lost ACKs:");
	respPayload = generateRandomPayload(10);
	String path = "test";
	int obs = 100;

	Request request = createRequest(GET, path, server);
	SynchronousNotificationListener notificationListener = new SynchronousNotificationListener(request);
	client.addNotificationListener(notificationListener);
	request.setObserve();
	client.sendRequest(request);

	server.expectRequest(CON, GET, path).storeMID("A").storeToken("B").observe(0).go();
	server.sendEmpty(ACK).loadMID("A").go();
	Thread.sleep(50);
	server.sendResponse(CON, CONTENT).loadToken("B").payload(respPayload).mid(++mid).observe(++obs).go();
	Thread.sleep(10);
	server.sendRequest(CON, GET, new Token(new byte[] {1,1}), mid).go();
	server.expectEmpty(ACK, mid).go();
	server.expectEmpty(ACK, mid).go();
	
	Thread.sleep(1000);
}
```

**Without any endpoint context check at all (long time ago) :**
0081: CON [MID=52852, T=836EBDB1714EE02C], GET, /test, observe(0)    ----->
0089: <-----   ACK [MID=52852]
0147: <-----   CON [MID=8001, T=836EBDB1714EE02C], 2.05, observe(101)
0152: ACK [MID=8001]   ----->
0159: ERROR: Server sent CON-GET    MID= 8001, Token=0101, OptionSet={}, no payload
0169: RST [MID=8001]   ----->

RST is triggered by 
```
11:58:48.312 WARN [UdpMatcher]: error receiving request CON-GET    MID= 8001, Token=0101, OptionSet={}, no payload again! (o.e.c.c.n.UdpMatcher$1.run:240)
java.lang.IllegalStateException: Exchange[L1, complete] retransmit on local exchange not allowed!
	at org.eclipse.californium.core.network.Exchange.retransmitResponse(Exchange.java:705)
	at org.eclipse.californium.core.network.stack.ReliabilityLayer.receiveRequest(ReliabilityLayer.java:225)
	at org.eclipse.californium.core.network.stack.AbstractLayer.receiveRequest(AbstractLayer.java:84)
	at org.eclipse.californium.core.network.stack.BaseCoapStack.receiveRequest(BaseCoapStack.java:133)
	at org.eclipse.californium.core.network.CoapEndpoint$1.receiveRequest(CoapEndpoint.java:250)
	at org.eclipse.californium.core.network.UdpMatcher$1.run(UdpMatcher.java:238)
	at org.eclipse.californium.elements.util.SerialExecutor$1.run(SerialExecutor.java:276)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
```

**Just before this PR :**
0083: CON [MID=64170, T=2390F59C8B943ED8], GET, /test, observe(0)    ----->
0090: <-----   ACK [MID=64170]
0151: <-----   CON [MID=8001, T=2390F59C8B943ED8], 2.05, observe(101)
0156: ACK [MID=8001]   ----->
0163: ERROR: Server sent CON-GET    MID= 8001, Token=0101, OptionSet={}, no payload

No message sent because of : 
```
12:13:45.790 ERROR [CoapEndpoint]: exception in protocol stage thread: null (o.e.c.c.n.CoapEndpoint$7.run:1042)
java.lang.NullPointerException: null
	at org.eclipse.californium.elements.KeySetEndpointContextMatcher.internalMatch(KeySetEndpointContextMatcher.java:102)
	at org.eclipse.californium.elements.KeySetEndpointContextMatcher.isToBeSent(KeySetEndpointContextMatcher.java:98)
	at org.eclipse.californium.core.network.UdpMatcher.receiveRequest(UdpMatcher.java:219)
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl.receiveRequest(CoapEndpoint.java:896)
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl.receiveMessage(CoapEndpoint.java:832)
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl.access$0(CoapEndpoint.java:823)
	at org.eclipse.californium.core.network.CoapEndpoint$InboxImpl$1.run(CoapEndpoint.java:811)
	at org.eclipse.californium.core.network.CoapEndpoint$7.run(CoapEndpoint.java:1040)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:473)
	at java.util.concurrent.FutureTask.run(FutureTask.java:262)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:178)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:292)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1152)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:622)
	at java.lang.Thread.run(Thread.java:748)
````

**With this PR**
0101: CON [MID=49028, T=6B26C34488253B90], GET, /test, observe(0)    ----->
0108: <-----   ACK [MID=49028]
0160: <-----   CON [MID=8001, T=6B26C34488253B90], 2.05, observe(101)
0161: ACK [MID=8001]   ----->
0171: ERROR: Server sent CON-GET    MID= 8001, Token=0101, OptionSet={}, no payload
0180: ERROR: Server received ACK-2.05   MID= 8001, Token=0101, OptionSet={"Content-Format":"text/plain", "Block2":"(szx=0/16, m=true, num=0)", "Size2":448}, "****************"

client is not "punished". we give the priority to the last message in this case.
**If we keep this behavior could we at least log with a warn that the device does not behave correctly ?**

**With this PR + source test** like this 
```java
EndpointContext sourceContext = request.getSourceContext();
Request previousRequest = previous.getCurrentRequest();
if (previous.getOrigin() == Origin.LOCAL) {
	EndpointContext previousSourceContext = previousRequest.getDestinationContext();
	// the previous response would be send with its previous context
	// using the current request context as connection context
	duplicate = endpointContextMatcher.isToBeSent(previousSourceContext, sourceContext);
} else {
	EndpointContext previousSourceContext = previousRequest.getSourceContext();
	// the previous response would be send with its previous context
	// using the current request context as connection context
	duplicate = endpointContextMatcher.isToBeSent(previousSourceContext, sourceContext);
}
```
0079: CON [MID=25149, T=1DBD9E5A0D5FFA14], GET, /test, observe(0)    ----->
0086: <-----   ACK [MID=25149]
0138: <-----   CON [MID=8001, T=1DBD9E5A0D5FFA14], 2.05, observe(101)
0140: ACK [MID=8001]   ----->
0152: ERROR: Server sent CON-GET    MID= 8001, Token=0101, OptionSet={}, no payload
0164: RST [MID=8001]   ----->
We get again the initial behavior with RST message.